ssl/t1_lib.c

```diff
@@ -2291,6 +2308,24 @@ int tls_choose_sigalg(SSL *s, int fatalerrs)
         int curve = -1, skip_ec = 0;
 #endif
 
+        /*
+         * Look for a certificate indicated by signature_algorithms_cert,
+         * if present.
+         */
+        if (s->s3->tmp.peer_cert_sigalgs != NULL) {
+            for (i = 0; i < s->s3->tmp.peer_cert_sigalgslen; i++) {
+                lu = tls1_lookup_sigalg(s->s3->tmp.peer_cert_sigalgs[i]);
+                if (lu == NULL || !ssl_has_cert(s, lu->sig_idx))
+                    continue;
+                s->s3->tmp.cert = &s->cert->pkeys[sig_idx];
+                s->cert->key = s->s3->tmp.cert;
+                break;
+            }
+            if (s->s3->tmp.cert == NULL)
+                SSLfatal(s, SSL_AD_HANDSHAKE_FAILURE, SSL_F_TLS_CHOOSE_SIGALG,
+                         SSL_R_NO_SUITABLE_SIGNATURE_ALGORITHM);
+                return 0;
```

Comment:
Err...shouldn't this all be in `{}`? Even if you do that though I'm not sure why we then fall through to the "Look for a certificate matching shared sigalgs" code below. That seems to do a lot of work setting up `sig_idx` which we then ignore because `s->s3->tmp.cert != NULL`. Or did I misunderstand something?

Reply:
We need to (re)compute ```lu``` using the signature_algorithms (non-_cert) extension, in order to set ```s->s3->tmp.sigalg``` to the proper value for signing TLS messages.  signature_algorithms_cert is *only* used for choosing what certificate (chain, but this currently doesn't look at the chain) to present.

Reply:
+1 to mattcaswell; missing braces.